Document the Ubuntu snapshot service (#375)
canonical/ubuntu-server-documentation: how-to/software/snapshot-service.md

(snapshot-service)=
# How to use the Ubuntu snapshot service

The Ubuntu snapshot service allows you to easily access old packages in the
Ubuntu archive based on past dates. Some of the use cases for this service
include

* Installing a superseded version of a package as at a particular date and

>   time, which is useful to troubleshoot bugs or regressions or to allow

Reviewer: ```suggestion
  time, which is useful to troubleshoot bugs or regressions, or to allow
```

  reproducibility.
* Setting upgrades to known states, which can be useful to ensure homogeneity
  for a fleet of Ubuntu installations and to ensure predictability upon package
  upgrades (including unattended upgrades).
* Support a structured update workflow by validating snapshots in different
  environments or by rolling out package upgrades in stages for a large set of
  machines.

This how-to guide shows how to set-up and use the snapshot service.

## Prerequisites

Snapshots are supported in Ubuntu 23.10 onwards and on updated installations of
Ubuntu 20.04 LTS (starting from `apt` 2.0.10) and Ubuntu 22.04 LTS (starting
from `apt` 2.4.11).

> In Ubuntu 24.04 LTS and later, `apt` will automatically detect whether a

Reviewer 2: suggestion (non-blocking): Consider adding a table of the different supported versions and their remarks.

Reviewer: +1 I think a small table will help provide clarity. I'm not sure what "updated installations of Ubuntu 20.04 and 22.04" means - what's the thing that needs to be updated?

```
deb http://archive.ubuntu.com/ubuntu jammy main universe restricted multiverse
deb http://archive.ubuntu.com/ubuntu jammy-updates main universe restricted multiverse
deb http://archive.ubuntu.com/ubuntu jammy-backports main restricted universe multiverse
deb http://security.ubuntu.com/ubuntu jammy-security main restricted universe multiverse
```

you can enable the snapshot service for all components in all pockets by changing it to

```
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu jammy main universe restricted multiverse
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu jammy-updates main universe restricted multiverse
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu jammy-backports main restricted universe multiverse
deb [snapshot=yes] http://security.ubuntu.com/ubuntu jammy-security main restricted universe multiverse
```

Without the setup shown above, `apt` will simply ignore snapshot related
command line options in Ubuntu releases lower than 24.04.

## Using the snapshot service

There are several options when it comes to setting up Ubuntu snapshot services.
Here we present the most relevant, supported ones.

> Below, we refer to the Snapshot ID as `$SNAPSHOT_ID`. This is a UTC date and

Reviewer: ```suggestion
Below, we refer to the snapshot ID as `$SNAPSHOT_ID`. This is a UTC date and
```

time formatted as YYYYMMDDTHHMMSSZ, e.g., 20240431T214500Z, which refers to the

> snapshot that represents the state of the Ubuntu archive on April 31st, 2024,

Reviewer 2: todo: There is no April 31
```suggestion
time formatted as YYYYMMDDTHHMMSSZ, e.g., 20240430T214500Z, which refers to the
snapshot that represents the state of the Ubuntu archive on April 30 2024,
```

Author: D'oh! 
Thanks :)

at 21:45:00 UTC.

Currently, the Ubuntu snapshot service provides snapshots for any date and time
**after 1 March 2023**.

### The `--snapshot` CLI option

The snapshot service can be used through the `--snapshot` (or `-S`) CLI option

> for `apt`, which receives a Snapshot ID argument.

Reviewer: is there a reason why the snapshot in "Snapshot ID" is capitalized? I found an instance earlier and thought it was a typo. Is there a specific convention? It confused me that most instances of snapshot are not capitalized but these are 🙃

```
apt update --snapshot $SNAPSHOT_ID
apt install $PACKAGE_NAME --snapshot $SNAPSHOT_ID
```

> you can also use the short, atomic version to merge the two actions above in a

Reviewer 2: typo:
```suggestion
You can also use the short, atomic version to merge the two actions above in a
```

single command, which will download the package information from the
repository, download, and install the package:

```
apt install $PACKAGE_NAME --update --snapshot $SNAPSHOT_ID
```

or using the short version of the snapshot option:

```

> apt install $PACKAGE_NAME --update --S $SNAPSHOT_ID

Reviewer 2: todo: Fix short version with single dash `-S` and consider just showing the short version, to avoid duplication. The `--snapshot` variant is already displayed above.
```suggestion
apt install $PACKAGE_NAME --update -S $SNAPSHOT_ID
```

Author: OK then, how about I leave this one as is (to show the usage of the short version of the command) and replace any other `-S` occurrences to `--snapshot`

  Installed: (none)
  Candidate: 24.0.7-0ubuntu4
  Version table:
     24.0.7-0ubuntu4 500
        500 https://snapshot.ubuntu.com/ubuntu/20240501T120000Z noble/universe amd64 Packages
```

If you forget to use the snapshot option, `apt` will operate over the regular archive:

```
# apt policy docker.io
docker.io:
  Installed: (none)
  Candidate: 28.2.2-0ubuntu1~24.04.1
  Version table:
     28.2.2-0ubuntu1~24.04.1 500
        500 http://archive.ubuntu.com/ubuntu noble-updates/universe amd64 Packages
     27.5.1-0ubuntu3~24.04.2 500
        500 http://security.ubuntu.com/ubuntu noble-security/universe amd64 Packages
     24.0.7-0ubuntu4 500
        500 http://archive.ubuntu.com/ubuntu noble/universe amd64 Packages
```

Then, you can install Docker from that snapshot:

```
# apt install docker.io --snapshot 20240501T120000Z
...
# docker --version
Docker version 24.0.7, build 24.0.7-0ubuntu4
```

Now, let's say you checked the [Docker package publishing history in
Launchpad](https://launchpad.net/ubuntu/+source/docker.io-app/+publishinghistory)
and (for some reason) decided you want to install version
`26.1.3-0ubuntu1~24.04.1`, which, as inferred from the publishing history,

> superseded `docker.io` `24.0.7-0ubuntu4.1` in noble-updates on `2024-11-25`.

Reviewer: ```suggestion
superseded `docker.io` `24.0.7-0ubuntu4.1` in `noble-updates` on `2024-11-25`.
```

> Then, `24.0.7-0ubuntu4.1` was removed from the updates pocket on `2024-11-26`.

Reviewer: ```suggestion
Then, `24.0.7-0ubuntu4.1` was removed from the `-updates` pocket on `2024-11-26`.
```

We could then use `20241126T230000Z` as the Snapshot ID to get the target
package (`26.1.3-0ubuntu1~24.04.1`):

```
# apt install docker.io --update -S 20241126T230000Z
...
# docker --version
Docker version 26.1.3, build 26.1.3-0ubuntu1~24.04.1
```

> **Note that, as the moment of this writing, the version of docker.io in the

Reviewer 2: nitpick (non-blocking): Use a more common phrase.
```suggestion
**Note that, as the time of this writing, the version of docker.io in the
```

> noble-security pocket is `27.5.1-0ubuntu3~24.04.2`, i.e., the version used in

Reviewer: ```suggestion
`noble-security` pocket is `27.5.1-0ubuntu3~24.04.2`, i.e., the version used in
```

the example above may be affected by known vulnerabilities. When using the
snapshot service, do make sure to check the latest version available in your
Ubuntu release security pocket to make sure you are not installing a vulnerable
package.**

### Configuring the snapshot service for specific repositories

In your `apt` repositories configuration files, it is possible to specify
snapshots to be used for each specific repository.

> For Ubuntu 24.04 LTS and later, which uses the deb822 style by default, we add

Reviewer: Good call - I'm not sure what the deb822 style means :D

a `Snapshot` option to the relevant sources file in the following format:

```
Snapshot: $SNAPSHOT_ID
```

For Ubuntu series lower than 24.04, we change the value of the `snapshot`
option in your sources file, i.e., if you had the following entry in your `/etc/apt/sources.list` file

```
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu ...
````

you replate the `yes` value with the Snapshot ID:

```
deb [snapshot=$SNAPSHOT_ID] http://archive.ubuntu.com/ubuntu ...
````

Let's say we want to pin a specific snapshot for all the pockets in our Ubuntu
24.04 LTS server. Then, we add the `Snapshot` options to
`/etc/apt/sources.list.d/ubuntu.sources`:

```
Types: deb
URIs: http://archive.ubuntu.com/ubuntu
Suites: noble noble-updates noble-backports
Components: main universe restricted multiverse
Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
Snapshot: 20250530T223000Z

Types: deb
URIs: http://security.ubuntu.com/ubuntu
Suites: noble-security
Components: main universe restricted multiverse
Signed-By: /usr/share/keyrings/ubuntu-archive-keyring.gpg
Snapshot: 20250530T223000Z
```

> Then, we can simply run

Reviewer: I like the suggested replacement, it also neatly avoids use of the word "simply"

```
# apt install --update docker.io
```

to install the Docker version present in that snapshot.

```
# docker --version
Docker version 27.5.1, build 27.5.1-0ubuntu3~24.04.1
```

> Note that, when you configure a repository to use a Snapshot using the method

Reviewer 2: typo:
```suggestion
Note that, when you configure a repository to use a snapshot using the method
```

```
# apt install --update docker.io -S 20251013T23:00Z
...
Hit:5 https://snapshot.ubuntu.com/ubuntu/20250530T223000Z noble InRelease
Hit:6 https://snapshot.ubuntu.com/ubuntu/20250530T223000Z noble-updates InRelease
Hit:7 https://snapshot.ubuntu.com/ubuntu/20250530T223000Z noble-backports InRelease
Hit:8 https://snapshot.ubuntu.com/ubuntu/20250530T223000Z noble-security InRelease
...
docker.io is already the newest version (27.5.1-0ubuntu3~24.04.1).
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
```

As you can see from the logs above, `apt` is still using the snapshot configure
in `/etc/apt/sources.list.d/ubuntu.sources`, ignoring the `--snapshot` CLI
option.

### Configuring the snapshot service globally

As an alternative to configuring a specific snapshot for each individual
repository, as described in the previous section, you can configure a snapshot
globally for all repositories that have snapshots enabled.

**If you are following this examples until here, make sure to revert the
changes to the sources configuration file made in the previous section**

Let's configure `apt` to default to a specific snapshot by setting the
`APT::Snapshot` option:

```

> # echo 'APT::Snapshot "20250801T111111Z";' | tee /etc/apt/apt.conf.d/50snapshot

Reviewer: I thought I'd set up the copybutton extension to strip the prompts, but you're right - it's not working. Leave that with me, I might need to add some regex to the configuration.

deb http://archive.ubuntu.com/ubuntu jammy-backports main restricted universe multiverse
deb http://security.ubuntu.com/ubuntu jammy-security main restricted universe multiverse
```

you can enable the snapshot service for [all components in all
pockets](https://documentation.ubuntu.com/project/how-ubuntu-is-made/concepts/package-archive/#pockets)
by changing it to

```
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu jammy main universe restricted multiverse
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu jammy-updates main universe restricted multiverse
deb [snapshot=yes] http://archive.ubuntu.com/ubuntu jammy-backports main restricted universe multiverse
deb [snapshot=yes] http://security.ubuntu.com/ubuntu jammy-security main restricted universe multiverse
```

Without the setup shown above, `apt` will simply ignore snapshot related
command line options in Ubuntu releases lower than 24.04.

The following table describes how the snapshot service can be used with each
supported Ubuntu release.

| Ubuntu Release       | Supports Snapshot Service?   | Needs to configure the `snapshot` option in the sources.list file? |

> |----------------------|------------------------------|--------------------------------------------------------------------|

Reviewer: ```suggestion
The following table describes whether the snapshot service can be used with each
supported Ubuntu release, and whether or not it is necessary to configure the `snapshot` option in the `sources.list` file.

| Ubuntu Release       | Supports Snapshot Service?   | Need to configure `sources.list`? |
|----------------------|------------------------------|--------------------------------------------------------------------|
```
